File: javascript/robot-name/robot-name.js
const LETTERS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const NUMBERS = '0123456789';
const usedNames = {};

const generateName = () => {
  let name = sampler(LETTERS, 2) + sampler(NUMBERS, 3);
  if (usedNames[name]) name = generateName();
  else usedNames[name] = true;
  return name;
};

const sampler = (possible, n) => {
  let output = '';
  for (let i = 0; i < n; i++)
    output += possible.charAt(Math.floor(Math.random() * possible.length));
  return output;
}

class Robot {
  constructor() {
    this.robotName = generateName()
  }

  get name() {
    return this.robotName;
  }

  reset() {
    this.robotName = generateName();
  }
}

if (module) module.exports = Robot;

File: ruby/robot-name/robot_name.rb
class Robot
  attr_reader :name
  @@names = {}

  def initialize
    generate_names
    @name = pick_name
  end

  def reset
    # mark name as used
    @name = pick_name
  end


  def self.forget; end

  private

  def pick_name

  end

  def generate_names
    letters = ('AA'..'ZZ').to_a
    numbers = (000..999).to_a

    #cycle through letters and numbers to get all possibilities
    #set name as unused

  end
end

#old code
# class Robot
#   attr_reader :name
#   @@used_names = {}

#   def initialize
#     @name = generate_name
#   end

#   def reset
#     @name = generate_name
#   end

#   def self.forget

#   end

#   private

#   def generate_name
#     letters = ('A'..'Z').to_a
#     numbers = (0..9).to_a
#     generated_name = letters.sample(2).join + numbers.sample(3).join.to_s
#     if @@used_names[generated_name]
#       generate_name
#     else
#       @@used_names[generated_name] = true
#       return generated_name
#     end
#   end
# end
